File: Languages/go/algorithms-dsa/leetcode/trees/invert_binary_tree.go
/*
LeetCode #226: Invert Binary Tree
Topic: Trees
Difficulty: Easy

Given the root of a binary tree, invert the tree (mirror it) and
return its root.

Approach: Recursive DFS. Swap left and right children at each node,
then recurse on both subtrees.

Time: O(n), Space: O(h) where h is height
*/
package main

import "fmt"

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func invertTree(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}
	root.Left, root.Right = root.Right, root.Left
	invertTree(root.Left)
	invertTree(root.Right)
	return root
}

func main() {
	tree := &TreeNode{4,
		&TreeNode{2, &TreeNode{1, nil, nil}, &TreeNode{3, nil, nil}},
		&TreeNode{7, &TreeNode{6, nil, nil}, &TreeNode{9, nil, nil}},
	}
	result := invertTree(tree)
	if result.Left.Val != 7 || result.Right.Val != 2 {
		fmt.Println("FAIL Test 1")
	} else {
		fmt.Println("PASS Test 1")
	}

	if invertTree(nil) != nil {
		fmt.Println("FAIL Test 2")
	} else {
		fmt.Println("PASS Test 2")
	}
}

File: Languages/cpp/algorithms-dsa/leetcode/trees/invert_binary_tree.cpp
/**
 * LeetCode 226: Invert Binary Tree
 * Topic: Trees
 * Difficulty: Easy
 *
 * Invert a binary tree (mirror it).
 * Time: O(n), Space: O(h) where h = height
 */
#include <iostream>
#include <cassert>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

TreeNode* invertTree(TreeNode* root) {
    if (!root) return nullptr;
    swap(root->left, root->right);
    invertTree(root->left);
    invertTree(root->right);
    return root;
}

int main() {
    // Build: [4, 2, 7, 1, 3, 6, 9]
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(7);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);
    root->right->left = new TreeNode(6);
    root->right->right = new TreeNode(9);

    invertTree(root);

    // After invert: root->left should be 7, root->right should be 2
    assert(root->left->val == 7);
    assert(root->right->val == 2);
    assert(root->left->left->val == 9);
    assert(root->right->right->val == 1);

    cout << "All tests passed!" << endl;
    return 0;
}
